### DISCUSSION TITLE
Figure caption not showing when using inline SVG

### DISCUSSION THREAD

**Discussion by netw0rkf10w** [quarto-dev/quarto-cli#5266] [Q&A] [answered]
Hello,

I'm using [this filter](https://github.com/pandoc-ext/diagram/blob/main/diagram.lua) for converting TikZ code to SVG in Quarto. When using the filter as is, it works (an example can be found at https://raw.githubusercontent.com/pandoc-ext/diagram/main/test/input-tikz.md). However, when I changed the code to replace the SVG file with inline SVG, the caption is not showing (I checked the output HTML and the `figure` container was not even generated).

More precisely, in the original code, an SVG image is created and store in `img`, then:

```lua
-- Use the block's filename attribute or create a new name by hashing the image content.
  local basename, _extension = pandoc.path.split_extension(
    props.filename or pandoc.sha1(img)
  )
  local fname = basename .. '.' .. extension_for_mimetype[imgtype]
  -- Store the data in the media bag:
  pandoc.mediabag.insert(fname, imgtype, img)
  -- Create a figure that contains just this image.
  local img_obj = pandoc.Image(props.alt, fname, "", props['image-attr'])
  return pandoc.Figure(pandoc.Plain{img_obj}, props.caption, props['fig-attr'])
```

What I did is, instead of creating a file in the media bag, I used directly the SVG as an inline HTML element:
```lua
return pandoc.Figure(pandoc.RawInline('html', img), props.caption, props['fig-attr'])
```
(the above line replaces the entire previous code block)

**Using Pandoc alone, it worked:**

```
quarto pandoc --lua-filter filters/diagram.lua -s tikz.qmd  -o tikz.html
```

However, when rendering `tikz.qmd` containing
```
filters:
  - diagram.lua
```
with Quarto, the SVG is displayed but not in a figure container and thus there's no caption.

Could you please help? Thank you very much in advance!

**Comment by mcanouil** [comment]
Could you share a more actionable example, such as a complete and small Quarto document using your filter?

**Reply by netw0rkf10w** [reply]
@mcanouil Sure! I tested the filter on the following `tikz.qmd` file (please change the extension to `.qmd`):

[tikz.txt](https://github.com/quarto-dev/quarto-cli/files/11299131/tikz.txt)

**Reply by mcanouil** [reply]
I can't render anything[^1], could you share a Git repository with everything?
Also which operating system and which version of Quarto are you using?

[^1]: The repository had many updates (<https://github.com/pandoc-ext/diagram>), especially the Lua filter and I have no way to know exactly what you have.

**Reply by netw0rkf10w** [reply]
@mcanouil Please check the full code here: https://github.com/netw0rkf10w/quarto-test. Please note that the filter requires `inkscape`.

The figures are in `about.qmd`. Doing `quarto preview`, one should see TikZ figures with captions on the About page.

Now, changing to inline SVG mode by setting `use_inline_svg = true` at [this line](https://github.com/netw0rkf10w/quarto-test/blob/1c52ed33179f229aa5e90290d86be2cd285bc644/filters/diagram.lua#L458), the figures will be still there but with no caption.

My quarto version is `1.3.326`, installed on macOS 10.15.7.

Thanks a lot!

**Reply by mcanouil** [reply]
I think the issue is that the figure environment is not handled by Quarto but by the Lua filter itself in this case.
I know @tarleb is working on this particular filter at the moment and some other filters which can include the corresponding Quarto extension (_e.g._, <https://github.com/pandoc-ext/section-bibliographies>).

**Reply by tarleb** [reply]
I believe the issue might be that Quarto currently uses the pandoc 2.* representation for figures. This is going to change AFAIK, but it might affect figure processing in current versions.

Feel free to raise this as an issue in the diagram repository: inline SVGs are something that I'd like to support out of the box.

**Reply by mcanouil** [reply]
Thanks @tarleb for your input, I'll open an issue later to track this (I need to check the code and/or with the team).
It is likely to be in 1.4 if indeed updating code related to figure representation (AST) is needed.

**Reply by netw0rkf10w** [reply]
@tarleb Thanks. Is there a workaround please? Also I couldn't find the diagram repo.

**Reply by mcanouil** [reply]
@netw0rkf10w I wrote the repository in one of my answer[^1], see <https://github.com/pandoc-ext/diagram>.

[^1]: <https://github.com/quarto-dev/quarto-cli/discussions/5266#discussioncomment-5692544>

**Reply by netw0rkf10w** [reply]
@mcanouil Sorry I thought @tarleb was referring to a repo of Quarto. Obviously I knew about the pandoc-ext repo because that's where I took the filter from. And I'm afraid that it is not appropriate to raise an issue there, because this issue is specific to Quarto and not Pandoc in general. As I mentioned in my first post, **the filter works just fine with Pandoc:** when doing `quarto pandoc --lua-filter filters/diagram.lua -s tikz.qmd  -o tikz.html`, it works! Thus, neither Pandoc nor the filter itself is the problem I guess.

**Reply by mcanouil** [reply]
It seems so as I said, it's likely an issue on how figure environment are handled within Quarto which might not be completely compatible with the filter at this moment and likely to be something for 1.4.

**Reply by tarleb** [reply, marked as answer]
Workaround: use the original `diagram.lua` filter and run the below filter *after* the Quarto filters:
``` lua
function Image (img)
  if img.src:match '%.svg$' then
    local mime_type, contents = pandoc.mediabag.lookup(img.src)
    if mime_type == 'image/svg+xml' and contents then
      return pandoc.RawInline('html', contents:gsub('^.*(%<svg%f[%s])', '%1'))
    end
  end
end
```

The time at which quarto filters are run can be controlled by using the special `quarto` filter:

``` yaml
filters:
  - diagram.lua
  - quarto
  - inline-svg.lua
```

(Untested, so might need some tweeking.)

**Reply by mcanouil** [reply]
I just tried it and it worked.

**Reply by mcanouil** [reply]
I opened #5297 to track this further in the future.

**Reply by netw0rkf10w** [reply]
@tarleb It works like a charm, thanks a lot! I've just realised that you are actually the author of `diagram.lua` (not sure why I didn't notice this before). Thanks for the filter!
Thanks to @mcanouil as well for the discussion!
